### DISCUSSION TITLE
BonVision: Logging when stimulus is rendered

### DISCUSSION THREAD

**Discussion by RobertLamprecht99** [bonsai-rx/bonsai#1203] [Q&A]
Hello, my work flow consists of a grating series with the parameters loaded in from a CSV file. I would like to record a time stamp pf when each successive grating is presented. Based on the information from https://bonvision.github.io//pages/06-Logging/#, I understand that a framed event logger must be placed after the render node. I am a bit confused as to how to place the logEvent node to record stimulus presentation. 

 I have attached screenshots of the work flow.
![workflow](https://user-images.githubusercontent.com/120683927/209032129-82a35265-c99a-401f-afba-93eb8456b1e1.PNG)

**Comment by bruno-f-cruz** [comment]
I think you might want to use the `FrameEventLogger` in this case. Regardless, these nodes will take any input convert it to its `String` representation and pass it along to a logger subject. The only thing you have to tell them is when/what to log. 
In your picture, `LogEvent` will be logging every sample that comes from `Draw`. If you only want to log the start of each presentation, you could have a second subscription to `Draw` inside your `CreateObservable`. Something like: `Draw->Take(1)->FrameEventLogger`.
Completely unrelated, but you should try to use `Timer(Shaders)` when working with shaders.

For future reference, include the workflow so we can edit the code directly :)

Hope it helps,
Bruno

**Reply by RobertLamprecht99** [reply]
Hello, thank you for the response. I am running into an error when I use the Draw->Take(1)->FrameEventLogger stream. I get the error that 'TimeStep' is not a member of the type 'OpenTK.Matrix4'
[GratingTrial.zip](https://github.com/bonsai-rx/bonsai/files/10288205/GratingTrial.zip)

**Reply by bruno-f-cruz** [reply]
Sorry, I misspoke...
`FrameEventLogger` actually instantiates the logger,  you will still use the `LogEvent` to log single events. Here's the motif:

![image](https://user-images.githubusercontent.com/7049351/209171712-39a39f46-3ce9-44d2-92da-2f40a94480cb.png)

**Reply by RobertLamprecht99** [reply]
Thank you, I've gotten it working. It's outputting a CSV with the exact time the first frame of each stimulus is rendering. Additionally, I am using the subscribe when node to delay rendering of the first grating until the amplifier data is received. Would this mean that the logger instantiated by the FrameEventLogger, is synced with the ephys data?

**Reply by bruno-f-cruz** [reply]
Kinda.. The logger starts with the first sample that the FPGA produces. But the alignment is not guaranteed with a lot of precision. If you want to align precisely to the ephys signal I would probably just put a photodiode on the screen that measures the onset of relevant frames.

**Reply by RobertLamprecht99** [reply]
I wanted to follow up and ask if there was an example work flow available that incorporated a photodiode.

**Reply by bruno-f-cruz** [reply]
If you want to flash a quad on the screen on every frame, you could do something like:
Where:
```
Int is set to 1
Mod is set to 2
```

![image](https://user-images.githubusercontent.com/7049351/219485931-90beaeb3-b750-4fcf-a961-87580cadffd9.png)

Then place the quad on top of your photodiode and you will know when each frame is rendered. I would probably add a `Timer` at the end to make sure the ephys acquisition has enough time to start so you don't lose any frames. You can also combine this strategy with the previous `LogEvent ` approach and save what was the color of the quad on every single frame. Since you know when each frame was toggled in the shader window, and on the photodiode that you are recording with the ephys system, you should be able to effectively align the two streams.

Hope it helps!

**Reply by RobertLamprecht99** [reply]
So each time the quad flashes, the photodiode captures the new sample in which the frame was rendered? This can then be matched up to the frame recorded in the event logger file?

**Reply by RobertLamprecht99** [reply]
I am experiencing an issue in which the quad is simply flashing on the screen momentarily. I would prefer for it to stay and oscillate between the black and white so that the back ground for the photodiode is consistent. Additionally, I am noticing that occasionally at random intervals, bonsai is not rendering the quad at all leading to dropping frames. I've attached the current workflow to this post.
[VisualStim+Ephys03172023.zip](https://github.com/bonsai-rx/bonsai/files/11054322/VisualStim%2BEphys03172023.zip)

**Reply by bruno-f-cruz** [reply]
Hi,
Any specific reason for your quad drawing logic being inside the CreateObservable? That is likely the reason why it is not consistent across frames. Move it outside and it should just toggle on every frame.

**Reply by RobertLamprecht99** [reply]
I moved it outside of the create observable but it is still just flashing. When the gratings get to the end and it goes to the gray screen, the quad oscillates rather than just flashing

**Comment by bruno-f-cruz** [comment]
Can you share the modified workflow along with the output of the diode and
the logger? Thanks

**Comment by RobertLamprecht99** [comment]
[VisualStim+Ephys03172023.zip](https://github.com/bonsai-rx/bonsai/files/11055121/VisualStim%2BEphys03172023.zip)
[framedevent2023-03-23T15_10_10.csv](https://github.com/bonsai-rx/bonsai/files/11055122/framedevent2023-03-23T15_10_10.csv)
[photodiode2023-03-23T15_10_10.zip](https://github.com/bonsai-rx/bonsai/files/11055128/photodiode2023-03-23T15_10_10.zip)

**Reply by bruno-f-cruz** [reply]
The FrameEventLogger would be the logger you want to look at since it tells you the difference between each frame. Additionally, please make sure to include the necessary information to parse the binary file. I think I got there by trial an error but it would save alot of time to the people trying to help.

So, if you look at the diode here's what you have:
![image](https://user-images.githubusercontent.com/7049351/227330012-2a321890-72ca-462b-95ca-f04848867199.png)

If you take the difference between each peak in photodiode samples you get:
![image](https://user-images.githubusercontent.com/7049351/227333250-a6cad9bb-118e-487c-94d4-2bd3df59935b.png)

So it looks like your quad is turning on-off-on every 8 photodiode cycles. Looking at the photodiode I can else tell you are probably using a projector with a color wheel since it looks like the off phase 2x as long as "on" phase (which makes sense since seem to be toggling between red and black). You should probably use "blue" instead of "red" since projectors tend to render BGR not RGB.

Anyway, looking at the output of each "On" it shows that the time between each rising edge is remarkably constant, hence I would assume that the quad is toggling at the frequency it should (i.e. on and off every frame). Just to be sure, I would record the same experiment but always keeping the quad on, just to validate the method.

When you mentioned: "I moved it outside of the create observable but it is still just flashing. When the gratings get to the end and it goes to the gray screen, the quad oscillates rather than just flashing". Was this just from visual inspection or did you actually see it in the photodiode output?
